File: packages/temper-placer/src/temper_placer/router_v6/quality/corridor.py
"""
Router V6 Quality: Corridor Consolidation + Track-Spread (U3)

Defines and computes:
- corridor_consolidation_score: ratio of co-routed track pairs to total pairs
  sharing a channel. Range 0.0-1.0. Higher is better.
- track_spread_score: ratio of max gap between adjacent tracks to target
  spacing. Range >= 1.0. 1.0 = optimal spacing.

Gate: corridor >= 0.7, track-spread <= 1.5 (provisional).

Part of temper-7rqf (Stage 6 - Quality Gate)
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from temper_placer.io._kicad_types import ParseResult


# ---------------------------------------------------------------------------
# Data types
# ---------------------------------------------------------------------------


@dataclass
class Channel:
    """A rectangular gap region between two component courtyards.

    A channel is a space between components where tracks can run,
    wider than ``3 * (track_width + min_clearance)``.
    """

    x_min: float
    y_min: float
    x_max: float
    y_max: float
    gap_width_mm: float  # Width of the gap in mm
    axis: str  # "horizontal" or "vertical"
    component_a: str
    component_b: str

    @property
    def center(self) -> tuple[float, float]:
        return ((self.x_min + self.x_max) / 2, (self.y_min + self.y_max) / 2)

    @property
    def region(self) -> tuple[float, float, float, float]:
        return (self.x_min, self.y_min, self.x_max, self.y_max)


@dataclass
class TrackSegment:
    """A track segment passing through a channel."""

    net: str
    x: float
    y: float
    width_mm: float
    layer: str

    @property
    def left_edge(self) -> float:
        return self.x - self.width_mm / 2.0

    @property
    def right_edge(self) -> float:
        return self.x + self.width_mm / 2.0

    @property
    def bottom_edge(self) -> float:
        return self.y - self.width_mm / 2.0

    @property
    def top_edge(self) -> float:
        return self.y + self.width_mm / 2.0


# ---------------------------------------------------------------------------
# Default parameters
# ---------------------------------------------------------------------------

_DEFAULT_COURTYARD_CLEARANCE_MM = 0.25
_DEFAULT_TRACK_WIDTH_MM = 0.2
_DEFAULT_MIN_CLEARANCE_MM = 0.15
_CHANNEL_WIDTH_MULTIPLIER = 3.0


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def corridor_consolidation_score(
    routed_pcb_path: Path,
    courtyard_clearance_mm: float | None = None,
    track_width_mm: float | None = None,
    min_clearance_mm: float | None = None,
) -> float:
    """Ratio of co-routed track pairs to total pairs sharing a channel.

    A channel is the gap between two component courtyards wider than
    3 track widths. 'Co-routed' means two tracks adjacent with no
    foreign track interleaved.

    Range 0.0-1.0. Higher is better. Gate: >= 0.7.

    Args:
        routed_pcb_path: Path to the routed .kicad_pcb file.
        courtyard_clearance_mm: Clearance margin for courtyard expansion.
        track_width_mm: Typical track width in mm.
        min_clearance_mm: Minimum clearance from netclass SSOT.

    Returns:
        Consolidation score in [0.0, 1.0]. 1.0 if no channels have >= 2 tracks.
    """
    result = _parse_pcb(routed_pcb_path)
    return _compute_consolidation(result, courtyard_clearance_mm, track_width_mm, min_clearance_mm)


def track_spread_score(
    routed_pcb_path: Path,
    courtyard_clearance_mm: float | None = None,
    track_width_mm: float | None = None,
    min_clearance_mm: float | None = None,
) -> float:
    """Ratio of max gap between adjacent track edges to target spacing.

    Range >= 1.0. 1.0 = tracks at exactly target spacing.
    > 1.5 = too sparse. Gate: <= 1.5.

    Target spacing = min_clearance + track_width from netclass SSOT.

    Args:
        routed_pcb_path: Path to the routed .kicad_pcb file.
        courtyard_clearance_mm: Clearance margin for courtyard expansion.
        track_width_mm: Typical track width in mm.
        min_clearance_mm: Minimum clearance from netclass SSOT.

    Returns:
        Track-spread score >= 0.0. 0.0 if no tracks to spread.
    """
    result = _parse_pcb(routed_pcb_path)
    return _compute_spread(result, courtyard_clearance_mm, track_width_mm, min_clearance_mm)


# ---------------------------------------------------------------------------
# Internal: PCB parsing and geometry
# ---------------------------------------------------------------------------


def corridor_consolidation_from_parse(
    result: ParseResult,
    courtyard_clearance_mm: float | None = None,
    track_width_mm: float | None = None,
    min_clearance_mm: float | None = None,
) -> float:
    """Corridor consolidation score from a ParseResult (for reuse)."""
    return _compute_consolidation(result, courtyard_clearance_mm, track_width_mm, min_clearance_mm)


def track_spread_from_parse(
    result: ParseResult,
    courtyard_clearance_mm: float | None = None,
    track_width_mm: float | None = None,
    min_clearance_mm: float | None = None,
) -> float:
    """Track-spread score from a ParseResult (for reuse)."""
    return _compute_spread(result, courtyard_clearance_mm, track_width_mm, min_clearance_mm)


def _parse_pcb(pcb_path: Path) -> ParseResult:
    """Parse a KiCad PCB file."""
    from temper_placer.io.kicad_parser import parse_kicad_pcb

    return parse_kicad_pcb(Path(pcb_path))


def _compute_consolidation(
    result: ParseResult,
    courtyard_clearance_mm: float | None,
    track_width_mm: float | None,
    min_clearance_mm: float | None,
) -> float:
    """Compute corridor consolidation score from a ParseResult."""
    if courtyard_clearance_mm is None:
        courtyard_clearance_mm = _DEFAULT_COURTYARD_CLEARANCE_MM
    if track_width_mm is None:
        track_width_mm = _DEFAULT_TRACK_WIDTH_MM
    if min_clearance_mm is None:
        min_clearance_mm = _DEFAULT_MIN_CLEARANCE_MM

    channel_min_gap = _CHANNEL_WIDTH_MULTIPLIER * (track_width_mm + min_clearance_mm)
    courtyards = _compute_courtyards(result, courtyard_clearance_mm)
    channels = _identify_channels(courtyards, channel_min_gap)

    if not channels:
        return 1.0

    tracks_by_channel = _assign_tracks_to_channels(result, channels)
    total_pairs = 0
    co_routed_pairs = 0

    for ch in channels:
        channel_tracks = tracks_by_channel.get(id(ch), [])
        if len(channel_tracks) < 2:
            continue

        if ch.axis == "vertical":
            channel_tracks.sort(key=lambda t: t.x)
        else:
            channel_tracks.sort(key=lambda t: t.y)

        n = len(channel_tracks)
        total_pairs += n * (n - 1) // 2

        for i in range(n - 1):
            for j in range(i + 1, n):
                if j == i + 1:
                    co_routed_pairs += 1
                else:
                    intervening_nets = {track.net for track in channel_tracks[i + 1 : j]}
                    if len(intervening_nets) <= 1 and (
                        not intervening_nets or intervening_nets == {channel_tracks[i].net}
                    ):
                        co_routed_pairs += 1

    if total_pairs == 0:
        return 1.0

    return co_routed_pairs / total_pairs


def _compute_spread(
    result: ParseResult,
    courtyard_clearance_mm: float | None,
    track_width_mm: float | None,
    min_clearance_mm: float | None,
) -> float:
    """Compute track-spread score from a ParseResult."""
    if courtyard_clearance_mm is None:
        courtyard_clearance_mm = _DEFAULT_COURTYARD_CLEARANCE_MM
    if track_width_mm is None:
        track_width_mm = _DEFAULT_TRACK_WIDTH_MM
    if min_clearance_mm is None:
        min_clearance_mm = _DEFAULT_MIN_CLEARANCE_MM

    target_spacing_mm = track_width_mm + min_clearance_mm
    channel_min_gap = _CHANNEL_WIDTH_MULTIPLIER * (track_width_mm + min_clearance_mm)
    courtyards = _compute_courtyards(result, courtyard_clearance_mm)
    channels = _identify_channels(courtyards, channel_min_gap)

    if not channels:
        return 0.0

    tracks_by_channel = _assign_tracks_to_channels(result, channels)
    overall_max_gap_mm = 0.0
    any_tracks_found = False

    for ch in channels:
        channel_tracks = tracks_by_channel.get(id(ch), [])
        if len(channel_tracks) < 2:
            continue

        any_tracks_found = True

        if ch.axis == "vertical":
            channel_tracks.sort(key=lambda t: t.x)
        else:
            channel_tracks.sort(key=lambda t: t.y)

        for i in range(len(channel_tracks) - 1):
            if ch.axis == "vertical":
                gap = channel_tracks[i + 1].left_edge - channel_tracks[i].right_edge
            else:
                gap = channel_tracks[i + 1].bottom_edge - channel_tracks[i].top_edge
            if gap > overall_max_gap_mm:
                overall_max_gap_mm = gap

    if not any_tracks_found:
        return 0.0

    if target_spacing_mm <= 0.0:
        return 0.0

    return overall_max_gap_mm / target_spacing_mm


@dataclass
class _Courtyard:
    """Component courtyard (bbox + clearance margin)."""

    ref: str
    x_min: float
    y_min: float
    x_max: float
    y_max: float


def _compute_courtyards(
    result: ParseResult,
    clearance_mm: float,
) -> list[_Courtyard]:
    """Compute component courtyards from the parsed PCB.

    Each courtyard is the component bbox expanded by ``clearance_mm`` on all sides.
    """
    courtyards: list[_Courtyard] = []
    for comp in result.netlist.components:
        if comp.initial_position is None:
            continue
        cx, cy = comp.initial_position
        half_w = comp.width / 2.0 + clearance_mm
        half_h = comp.height / 2.0 + clearance_mm
        courtyards.append(
            _Courtyard(
                ref=comp.ref,
                x_min=cx - half_w,
                y_min=cy - half_h,
                x_max=cx + half_w,
                y_max=cy + half_h,
            )
        )
    return courtyards


def _identify_channels(
    courtyards: list[_Courtyard],
    min_gap_width_mm: float,
) -> list[Channel]:
    """Identify channels (gaps between courtyards wider than ``min_gap_width_mm``).

    A channel is formed where the projections of two courtyards overlap
    in one axis and the gap between them in the orthogonal axis is
    wider than ``min_gap_width_mm``.
    """
    if len(courtyards) < 2:
        return []

    channels: list[Channel] = []

    for i, ca in enumerate(courtyards):
        for j, cb in enumerate(courtyards):
            if j <= i:
                continue

            # Vertical channel: x-projections overlap, y-gap is wide enough
            x_overlap = _overlap(ca.x_min, ca.x_max, cb.x_min, cb.x_max)
            if x_overlap is not None:
                gap = _gap(ca.y_max, cb.y_min)
                if gap > min_gap_width_mm:
                    x0, x1 = x_overlap
                    if ca.y_max < cb.y_min:
                        channels.append(
                            Channel(
                                x_min=x0,
                                y_min=ca.y_max,
                                x_max=x1,
                                y_max=cb.y_min,
                                gap_width_mm=gap,
                                axis="vertical",
                                component_a=ca.ref,
                                component_b=cb.ref,
                            )
                        )
                    else:
                        channels.append(
                            Channel(
                                x_min=x0,
                                y_min=cb.y_max,
                                x_max=x1,
                                y_max=ca.y_min,
                                gap_width_mm=gap,
                                axis="vertical",
                                component_a=ca.ref,
                                component_b=cb.ref,
                            )
                        )

            # Horizontal channel: y-projections overlap, x-gap is wide enough
            y_overlap = _overlap(ca.y_min, ca.y_max, cb.y_min, cb.y_max)
            if y_overlap is not None:
                gap = _gap(ca.x_max, cb.x_min)
                if gap > min_gap_width_mm:
                    y0, y1 = y_overlap
                    if ca.x_max < cb.x_min:
                        channels.append(
                            Channel(
                                x_min=ca.x_max,
                                y_min=y0,
                                x_max=cb.x_min,
                                y_max=y1,
                                gap_width_mm=gap,
                                axis="horizontal",
                                component_a=ca.ref,
                                component_b=cb.ref,
                            )
                        )
                    else:
                        channels.append(
                            Channel(
                                x_min=cb.x_max,
                                y_min=y0,
                                x_max=ca.x_min,
                                y_max=y1,
                                gap_width_mm=gap,
                                axis="horizontal",
                                component_a=ca.ref,
                                component_b=cb.ref,
                            )
                        )

    return channels


def _assign_tracks_to_channels(
    result: ParseResult,
    channels: list[Channel],
) -> dict[int, list[TrackSegment]]:
    """Assign trace segments to the channels they pass through.

    Each track segment midpoint that falls within a channel's region
    is assigned to that channel.

    Returns:
        dict mapping ``id(channel)`` to list of ``TrackSegment``.
    """
    tracks: dict[int, list[TrackSegment]] = {id(ch): [] for ch in channels}

    for trace in result.traces:
        mid_x = (trace.start[0] + trace.end[0]) / 2.0
        mid_y = (trace.start[1] + trace.end[1]) / 2.0
        seg = TrackSegment(
            net=trace.net or "",
            x=mid_x,
            y=mid_y,
            width_mm=trace.width,
            layer=trace.layer,
        )
        for ch in channels:
            if _point_in_rect(mid_x, mid_y, ch.x_min, ch.y_min, ch.x_max, ch.y_max):
                tracks[id(ch)].append(seg)

    return tracks


# ---------------------------------------------------------------------------
# Geometry helpers
# ---------------------------------------------------------------------------


def _overlap(a_min: float, a_max: float, b_min: float, b_max: float) -> tuple[float, float] | None:
    """Return the overlapping interval of two ranges, or None."""
    o_min = max(a_min, b_min)
    o_max = min(a_max, b_max)
    if o_min < o_max:
        return (o_min, o_max)
    return None


def _gap(a_max: float, b_min: float) -> float:
    """Return the gap between two ranges (positive = separated, negative = overlapping)."""
    return b_min - a_max


def _point_in_rect(
    x: float, y: float, x_min: float, y_min: float, x_max: float, y_max: float
) -> bool:
    """Check if a point is inside a rectangle."""
    return x_min <= x <= x_max and y_min <= y <= y_max

File: packages/temper-quality-oracle/src/cluster_f/corridor.rs
//! Port of `temper_placer.router_v6.quality.corridor`.
//!
//! Two structural facts from the oracle header are reproduced deliberately:
//!
//! * **Courtyards are board-relative, tracks are page-absolute.** The kernel
//!   compares the two frames, so on real boards no track is ever assigned to a
//!   channel and both scores collapse to their empty-input constants. That is
//!   defect 2, reported and pinned, not repaired here.
//! * **Both `else` arms of `_identify_channels` are unreachable.** `gap` is
//!   `cb.min - ca.max` and the guard `gap > min_gap_width_mm` (with a positive
//!   threshold) already implies `ca.max < cb.min`, so the `if` always wins.
//!   They are ported anyway, because the threshold is an argument and a
//!   non-positive one reaches them.
//!
//! `_assign_tracks_to_channels` keys by `id(ch)` in Python — a CPython object
//! address, so two value-equal channels are distinct keys. The port keys by
//! **channel index**, which has the same distinctness and no address
//! dependence.

use super::board::ParseView;
use super::pyfloat::{py_max2, py_min2};

pub const DEFAULT_COURTYARD_CLEARANCE_MM: f64 = 0.25;
pub const DEFAULT_TRACK_WIDTH_MM: f64 = 0.2;
pub const DEFAULT_MIN_CLEARANCE_MM: f64 = 0.15;
pub const CHANNEL_WIDTH_MULTIPLIER: f64 = 3.0;

#[derive(Clone, Debug)]
pub struct Courtyard {
    pub reference: String,
    pub x_min: f64,
    pub y_min: f64,
    pub x_max: f64,
    pub y_max: f64,
}

#[derive(Clone, Debug)]
pub struct Channel {
    pub x_min: f64,
    pub y_min: f64,
    pub x_max: f64,
    pub y_max: f64,
    pub gap_width_mm: f64,
    pub axis: &'static str,
    pub component_a: String,
    pub component_b: String,
}

#[derive(Clone, Debug)]
pub struct TrackSegment {
    pub net: String,
    pub x: f64,
    pub y: f64,
    pub width_mm: f64,
    pub layer: String,
}

impl TrackSegment {
    /// `self.x - self.width_mm / 2.0` — divide, then subtract (catalog B7).
    #[inline]
    pub fn left_edge(&self) -> f64 {
        self.x - self.width_mm / 2.0
    }
    #[inline]
    pub fn right_edge(&self) -> f64 {
        self.x + self.width_mm / 2.0
    }
    #[inline]
    pub fn bottom_edge(&self) -> f64 {
        self.y - self.width_mm / 2.0
    }
    #[inline]
    pub fn top_edge(&self) -> f64 {
        self.y + self.width_mm / 2.0
    }
}

/// `_overlap` — `max(a_min, b_min)` / `min(a_max, b_max)` with CPython's
/// first-argument-on-NaN builtins, then a strict `<` so a touching pair is
/// `None`.
#[inline]
pub fn overlap(a_min: f64, a_max: f64, b_min: f64, b_max: f64) -> Option<(f64, f64)> {
    let o_min = py_max2(a_min, b_min);
    let o_max = py_min2(a_max, b_max);
    if o_min < o_max { Some((o_min, o_max)) } else { None }
}

/// `_gap` — `b_min - a_max`.
#[inline]
pub fn gap(a_max: f64, b_min: f64) -> f64 {
    b_min - a_max
}

/// `_point_in_rect` — the chained comparison, so any NaN is `False`.
#[inline]
pub fn point_in_rect(x: f64, y: f64, x_min: f64, y_min: f64, x_max: f64, y_max: f64) -> bool {
    x_min <= x && x <= x_max && y_min <= y && y <= y_max
}

/// `_compute_courtyards` — bbox expanded by `clearance_mm` on all sides.
///
/// The half-extent absorbs the clearance **before** the centre offset
/// (`half_w = comp.width / 2.0 + clearance_mm`), which is a different f64 from
/// offsetting and then expanding.
pub fn compute_courtyards(view: &ParseView, clearance_mm: f64) -> Vec<Courtyard> {
    let mut out = Vec::new();
    for comp in &view.components {
        let Some((cx, cy)) = comp.initial_position else {
            continue;
        };
        let half_w = comp.width / 2.0 + clearance_mm;
        let half_h = comp.height / 2.0 + clearance_mm;
        out.push(Courtyard {
            reference: comp.reference.clone(),
            x_min: cx - half_w,
            y_min: cy - half_h,
            x_max: cx + half_w,
            y_max: cy + half_h,
        });
    }
    out
}

/// `_identify_channels` — gaps between courtyards wider than the threshold.
pub fn identify_channels(courtyards: &[Courtyard], min_gap_width_mm: f64) -> Vec<Channel> {
    if courtyards.len() < 2 {
        return Vec::new();
    }
    let mut channels = Vec::new();
    for (i, ca) in courtyards.iter().enumerate() {
        for cb in courtyards.iter().skip(i + 1) {
            // Vertical channel: x-projections overlap, y-gap wide enough.
            if let Some((x0, x1)) = overlap(ca.x_min, ca.x_max, cb.x_min, cb.x_max) {
                let g = gap(ca.y_max, cb.y_min);
                if g > min_gap_width_mm {
                    let (y_min, y_max) = if ca.y_max < cb.y_min {
                        (ca.y_max, cb.y_min)
                    } else {
                        (cb.y_max, ca.y_min)
                    };
                    channels.push(Channel {
                        x_min: x0,
                        y_min,
                        x_max: x1,
                        y_max,
                        gap_width_mm: g,
                        axis: "vertical",
                        component_a: ca.reference.clone(),
                        component_b: cb.reference.clone(),
                    });
                }
            }
            // Horizontal channel: y-projections overlap, x-gap wide enough.
            if let Some((y0, y1)) = overlap(ca.y_min, ca.y_max, cb.y_min, cb.y_max) {
                let g = gap(ca.x_max, cb.x_min);
                if g > min_gap_width_mm {
                    let (x_min, x_max) = if ca.x_max < cb.x_min {
                        (ca.x_max, cb.x_min)
                    } else {
                        (cb.x_max, ca.x_min)
                    };
                    channels.push(Channel {
                        x_min,
                        y_min: y0,
                        x_max,
                        y_max: y1,
                        gap_width_mm: g,
                        axis: "horizontal",
                        component_a: ca.reference.clone(),
                        component_b: cb.reference.clone(),
                    });
                }
            }
        }
    }
    channels
}

/// `_assign_tracks_to_channels`, keyed by channel index instead of `id(ch)`.
///
/// A segment whose midpoint lands in several channels is appended to each,
/// exactly as the Python does with one shared `TrackSegment` object; the
/// kernels only ever sort those lists independently, so cloning is equivalent.
pub fn assign_tracks_to_channels(
    view: &ParseView,
    channels: &[Channel],
) -> Vec<Vec<TrackSegment>> {
    let mut tracks: Vec<Vec<TrackSegment>> = vec![Vec::new(); channels.len()];
    for trace in &view.traces {
        let mid_x = (trace.start.0.f() + trace.end.0.f()) / 2.0;
        let mid_y = (trace.start.1.f() + trace.end.1.f()) / 2.0;
        let seg = TrackSegment {
            net: trace.net.clone().unwrap_or_default(),
            x: mid_x,
            y: mid_y,
            width_mm: trace.width.f(),
            layer: trace.layer.clone(),
        };
        for (idx, ch) in channels.iter().enumerate() {
            if point_in_rect(mid_x, mid_y, ch.x_min, ch.y_min, ch.x_max, ch.y_max) {
                tracks[idx].push(seg.clone());
            }
        }
    }
    tracks
}

/// Stable sort on a bare float key, matching `list.sort(key=...)`.
///
/// Python's timsort is stable and only ever asks `a < b`; Rust's `sort_by` is
/// stable too, so the two agree for any key set on which `<` is a strict weak
/// ordering. A NaN key breaks transitivity and the resulting order becomes an
/// artifact of each language's merge strategy — see the reachability note in
/// the PR report; no cluster-F input reaches that state.
fn stable_sort_by_key(tracks: &mut [TrackSegment], vertical: bool) {
    tracks.sort_by(|a, b| {
        let (ka, kb) = if vertical { (a.x, b.x) } else { (a.y, b.y) };
        if ka < kb {
            std::cmp::Ordering::Less
        } else if kb < ka {
            std::cmp::Ordering::Greater
        } else {
            std::cmp::Ordering::Equal
        }
    });
}

fn resolve(value: Option<f64>, default: f64) -> f64 {
    value.unwrap_or(default)
}

/// `_compute_consolidation`.
///
/// The threshold is `3.0 * (track_width + min_clearance)` with the add grouped
/// first — `1.0499999999999998`, which is neither `1.05` nor
/// `3.0*a + 3.0*b`. The inner pair loop is quadratic and recomputes
/// `intervening_nets` per pair; that is behaviour-preserving but is left
/// un-reassociated so the arithmetic order stays identical.
pub fn compute_consolidation(
    view: &ParseView,
    courtyard_clearance_mm: Option<f64>,
    track_width_mm: Option<f64>,
    min_clearance_mm: Option<f64>,
) -> f64 {
    let courtyard_clearance_mm = resolve(courtyard_clearance_mm, DEFAULT_COURTYARD_CLEARANCE_MM);
    let track_width_mm = resolve(track_width_mm, DEFAULT_TRACK_WIDTH_MM);
    let min_clearance_mm = resolve(min_clearance_mm, DEFAULT_MIN_CLEARANCE_MM);

    let channel_min_gap = CHANNEL_WIDTH_MULTIPLIER * (track_width_mm + min_clearance_mm);
    let courtyards = compute_courtyards(view, courtyard_clearance_mm);
    let channels = identify_channels(&courtyards, channel_min_gap);
    if channels.is_empty() {
        return 1.0;
    }

    let mut tracks_by_channel = assign_tracks_to_channels(view, &channels);
    let mut total_pairs: i64 = 0;
    let mut co_routed_pairs: i64 = 0;

    for (idx, ch) in channels.iter().enumerate() {
        let channel_tracks = &mut tracks_by_channel[idx];
        if channel_tracks.len() < 2 {
            continue;
        }
        stable_sort_by_key(channel_tracks, ch.axis == "vertical");

        let n = channel_tracks.len();
        total_pairs += (n as i64) * (n as i64 - 1) / 2;

        for i in 0..n - 1 {
            for j in i + 1..n {
                if j == i + 1 {
                    co_routed_pairs += 1;
                } else {
                    let mut intervening: Vec<&str> = Vec::new();
                    for t in &channel_tracks[i + 1..j] {
                        if !intervening.contains(&t.net.as_str()) {
                            intervening.push(t.net.as_str());
                        }
                    }
                    if intervening.len() <= 1
                        && (intervening.is_empty()
                            || intervening[0] == channel_tracks[i].net.as_str())
                    {
                        co_routed_pairs += 1;
                    }
                }
            }
        }
    }

    if total_pairs == 0 {
        return 1.0;
    }
    co_routed_pairs as f64 / total_pairs as f64
}

/// `_compute_spread`.
pub fn compute_spread(
    view: &ParseView,
    courtyard_clearance_mm: Option<f64>,
    track_width_mm: Option<f64>,
    min_clearance_mm: Option<f64>,
) -> f64 {
    let courtyard_clearance_mm = resolve(courtyard_clearance_mm, DEFAULT_COURTYARD_CLEARANCE_MM);
    let track_width_mm = resolve(track_width_mm, DEFAULT_TRACK_WIDTH_MM);
    let min_clearance_mm = resolve(min_clearance_mm, DEFAULT_MIN_CLEARANCE_MM);

    let target_spacing_mm = track_width_mm + min_clearance_mm;
    let channel_min_gap = CHANNEL_WIDTH_MULTIPLIER * (track_width_mm + min_clearance_mm);
    let courtyards = compute_courtyards(view, courtyard_clearance_mm);
    let channels = identify_channels(&courtyards, channel_min_gap);
    if channels.is_empty() {
        return 0.0;
    }

    let mut tracks_by_channel = assign_tracks_to_channels(view, &channels);
    let mut overall_max_gap_mm = 0.0f64;
    let mut any_tracks_found = false;

    for (idx, ch) in channels.iter().enumerate() {
        let channel_tracks = &mut tracks_by_channel[idx];
        if channel_tracks.len() < 2 {
            continue;
        }
        any_tracks_found = true;
        let vertical = ch.axis == "vertical";
        stable_sort_by_key(channel_tracks, vertical);

        for i in 0..channel_tracks.len() - 1 {
            let g = if vertical {
                channel_tracks[i + 1].left_edge() - channel_tracks[i].right_edge()
            } else {
                channel_tracks[i + 1].bottom_edge() - channel_tracks[i].top_edge()
            };
            if g > overall_max_gap_mm {
                overall_max_gap_mm = g;
            }
        }
    }

    if !any_tracks_found {
        return 0.0;
    }
    if target_spacing_mm <= 0.0 {
        return 0.0;
    }
    overall_max_gap_mm / target_spacing_mm
}

#[cfg(any(test, feature = "wasm-registry"))]
#[allow(dead_code, unused_imports, clippy::unwrap_used, clippy::expect_used)]
pub(crate) mod tests {
    use super::*;

    #[cfg_attr(test, test)]
    fn overlap_is_none_when_ranges_merely_touch() {
        assert!(overlap(0.0, 10.0, 10.0, 20.0).is_none());
        assert_eq!(overlap(0.0, 10.0, 5.0, 15.0), Some((5.0, 10.0)));
        // Signed zeros: -0.0 < 0.0 is False.
        assert!(overlap(-0.0, 0.0, -0.0, 0.0).is_none());
        // NaN keeps the FIRST argument: max(0.0, NaN) is 0.0, not NaN, so the
        // pair still overlaps. `f64::max` would return 0.0 here too, but
        // max(NaN, 0.0) is where the two diverge.
        assert_eq!(overlap(0.0, 10.0, f64::NAN, 5.0), Some((0.0, 5.0)));
        assert!(overlap(f64::NAN, 10.0, 0.0, 5.0).is_none());
    }

    #[cfg_attr(test, test)]
    fn channel_threshold_is_the_grouped_expression() {
        let threshold = CHANNEL_WIDTH_MULTIPLIER * (DEFAULT_TRACK_WIDTH_MM + DEFAULT_MIN_CLEARANCE_MM);
        assert_ne!(threshold, 1.05);
        assert_ne!(
            threshold,
            CHANNEL_WIDTH_MULTIPLIER * DEFAULT_TRACK_WIDTH_MM
                + CHANNEL_WIDTH_MULTIPLIER * DEFAULT_MIN_CLEARANCE_MM
        );
        assert_eq!(threshold, 1.0499999999999998);
    }

    #[cfg_attr(test, test)]
    fn identify_channels_is_order_asymmetric() {
        let lower = Courtyard {
            reference: "U1".into(),
            x_min: 0.0,
            y_min: 0.0,
            x_max: 10.0,
            y_max: 0.0,
        };
        let upper = Courtyard {
            reference: "U2".into(),
            x_min: 0.0,
            y_min: 5.0,
            x_max: 10.0,
            y_max: 15.0,
        };
        let threshold = 1.0499999999999998;
        assert_eq!(
            identify_channels(&[lower.clone(), upper.clone()], threshold).len(),
            1
        );
        assert_eq!(identify_channels(&[upper, lower], threshold).len(), 0);
    }

    // --- BEGIN generated by scripts/gen_wasm_test_registry.py: tests ---
    /// Every `#[test]` in this module, as a callable the `wasm32`
    /// entry point can invoke by index.  Generated because these
    /// functions are private to this module and unreachable from
    /// anywhere a registry could otherwise live.
    pub const WASM_TESTS: &[(&str, fn())] = &[
        ("cluster_f::corridor::tests::overlap_is_none_when_ranges_merely_touch", overlap_is_none_when_ranges_merely_touch),
        ("cluster_f::corridor::tests::channel_threshold_is_the_grouped_expression", channel_threshold_is_the_grouped_expression),
        ("cluster_f::corridor::tests::identify_channels_is_order_asymmetric", identify_channels_is_order_asymmetric),
    ];
    // --- END generated by scripts/gen_wasm_test_registry.py: tests ---
}
